### PR TITLE
Bluetooth: l2cap: make TX path service all connections

### LINKED ISSUE #74427: Bluetooth: Host: TX path doesn't process all ready ACL connections
**Describe the bug**
The host doesn't send as much data as it has resources for.

**To Reproduce**
Run this test: https://github.com/zephyrproject-rtos/zephyr/pull/74401/commits/7950b44050a1e5151739a7492df3d21edf6f0844

-> See that not all the L2CAP channels (that are on separate ACLs) are not pulled from immediately as they should.

**Expected behavior**
The Host should send as much data as it can to the controller if it has resources to do so.

**Impact**
Annoyance. Higher data latency.

**Logs and console output**
image

**Additional context**
Discovered while fixing https://github.com/zephyrproject-rtos/zephyr/issues/74321 .
This regression is due to https://github.com/zephyrproject-rtos/zephyr/pull/72090 which rewrites the TX data path.

### PR DESCRIPTION
**Pull request by jori-nordic**
Similar to ISO connections, ACL connections are not serviced as fast as
possible. Change this, and try to send as much as we have resources for.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/74427

### CODE REVIEW

**Review by theob-pro — DISMISSED** [review]

File: tests/bsim/bluetooth/host/l2cap/many_conns/src/main.c [hunk 1]
```diff
@@ -0,0 +1,488 @@
...
+#include "common.h"
+
+#define LOG_MODULE_NAME main
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(LOG_MODULE_NAME, LOG_LEVEL_INF);
+
+CREATE_FLAG(is_connected);
+CREATE_FLAG(flag_l2cap_connected);
+
+#define __stronk
```
Comment (Thalley) [inline tests/bsim/bluetooth/host/l2cap/many_conns/src/main.c:19]: ... :D


```suggestion
```

File: subsys/bluetooth/host/l2cap.c [hunk 2]
```diff
@@ -853,14 +853,27 @@ static bool chan_has_credits(struct bt_l2cap_le_chan *lechan)
 #endif
 }
 
+__weak void bt_l2cap_data_pull_spy(struct bt_conn *conn,
+				   struct bt_l2cap_le_chan *lechan,
+				   size_t amount,
+				   size_t *length)
+{
+}
```
Comment (Thalley) [inline subsys/bluetooth/host/l2cap.c:861]: Suggest to incorperate `test` in the name of this function if it's specifically only for testing.

File: tests/bsim/bluetooth/host/l2cap/compile.sh [hunk 3]
```diff
@@ -10,6 +10,7 @@ set -ue
 
 source ${ZEPHYR_BASE}/tests/bsim/compile.source
 
+app=tests/bsim/bluetooth/host/l2cap/many_conns compile
```
Comment (Thalley) [inline tests/bsim/bluetooth/host/l2cap/compile.sh:13]: Surprised you didn't go with `many_conns_handle_it` :P

File: tests/bsim/bluetooth/host/l2cap/many_conns/src/main.c [hunk 4]
```diff
@@ -0,0 +1,488 @@
...
+	/* Verify SDU data matches TX'd data. */
+	int pos = memcmp(buf->data, tx_data, buf->len);
+
+	if (pos != 0) {
+		LOG_ERR("RX data doesn't match TX: pos %d", pos);
+		LOG_HEXDUMP_ERR(buf->data, buf->len, "RX data");
+		LOG_HEXDUMP_INF(tx_data, buf->len, "TX data");
+
+		for (uint16_t p = 0; p < buf->len; p++) {
+			__ASSERT(buf->data[p] == tx_data[p],
```
Comment (Thalley) [inline tests/bsim/bluetooth/host/l2cap/many_conns/src/main.c:106]: Why not 
```suggestion
			ASSERT(buf->data[p] == tx_data[p],
```

File: tests/bsim/bluetooth/host/l2cap/many_conns/src/main.c [hunk 5]
```diff
@@ -0,0 +1,488 @@
...
+static void connected(struct bt_conn *conn, uint8_t conn_err)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	if (conn_err) {
+		FAIL("Failed to connect to %s (%u)", addr, conn_err);
+		return;
+	}
```
Comment (Thalley) [inline tests/bsim/bluetooth/host/l2cap/many_conns/src/main.c:205]: `ASSERT`?

File: tests/bsim/bluetooth/host/l2cap/many_conns/src/main.c [hunk 6]
```diff
@@ -0,0 +1,488 @@
...
+	/* Prepare tx_data */
+	for (size_t i = 0; i < sizeof(tx_data); i++) {
+		tx_data[i] = (uint8_t)i;
+	}
+
+	err = bt_enable(NULL);
+	if (err) {
+		FAIL("Can't enable Bluetooth (err %d)", err);
+		return;
+	}
```
Comment (Thalley) [inline tests/bsim/bluetooth/host/l2cap/many_conns/src/main.c:261]: `ASSERT`?

File: tests/bsim/bluetooth/host/l2cap/many_conns/src/main.c [hunk 7]
```diff
@@ -0,0 +1,488 @@
...
+
+	LOG_DBG("Peripheral Bluetooth initialized.");
+	LOG_DBG("Connectable advertising...");
+	err = bt_le_adv_start(BT_LE_ADV_CONN_OT, NULL, 0, NULL, 0);
+	if (err) {
+		FAIL("Advertising failed to start (err %d)", err);
+		return;
+	}
```
Comment (Thalley) [inline tests/bsim/bluetooth/host/l2cap/many_conns/src/main.c:269]: `ASSERT`?

**Review by Thalley — CHANGES_REQUESTED** [review]
No memes please :)

File: tests/bsim/bluetooth/host/l2cap/many_conns/src/main.c (hunk 1 above)
Comment (jori-nordic) [reply tests/bsim/bluetooth/host/l2cap/many_conns/src/main.c:19]: on it :+1: 

download (2)

**Review by Thalley — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]

**Review by alwa-nordic — COMMENTED** [review]
I think this is not a correct fix. The better fix is to check `should_stop_tx` in `bt_conn_tx_processor` and `bt_tx_irq_raise` there.

I can't see the direct reason to raise a "I have work to do" flag from somewhere where there is no work to be done. Is this an optimization?

**Comment by jori-nordic** [comment]
> should_stop_tx

I prefer re-triggering the same code-path instead of having two different code-paths for pulling data / finding out we have data.

**Comment by alwa-nordic** [comment]
> > should_stop_tx
> 
> I prefer re-triggering the same code-path instead of having two different code-paths for pulling data / finding out we have data.

Ok, how about an unconditional `bt_tx_irq_raise` in `bt_conn_tx_processor` just before calling `tx_data_pull`?

**Comment by jori-nordic** [comment]
> We can't do this from conn.c as we don't know if the NULL returned is because the current channel is out of data or because it has data but it can't send it (e.g. the current buf is being "viewed" already).

From the corresponding ISO PR. TLDR we get an infinite loop if you can't send the data for some other reason. Try it.

**Comment by alwa-nordic** [comment]
Ok. Free to merge this. I trust you understand it better than I do right now. :+1:

**Pull request merged**